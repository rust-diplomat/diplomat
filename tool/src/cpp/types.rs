use core::fmt;
use diplomat_core::Env;

use diplomat_core::ast;

use crate::cpp::config::LibraryConfig;

#[derive(Debug, Clone, Copy)]
pub enum RefType {
    Owned,
    BorrowedNullable,
    BorrowedNonNull,
}

pub fn gen_type(
    typ: &ast::TypeName,
    in_path: &ast::Path,
    ref_type: Option<RefType>,
    env: &Env,
    library_config: &LibraryConfig,
    // whether we are generating a struct field.
    // structs shouldn't have `const` fields, otherwise we lose
    // our assignment operators
    in_struct: bool,
) -> Result<String, fmt::Error> {
    let mut s = String::new();
    gen_type_inner(
        typ,
        in_path,
        ref_type,
        env,
        library_config,
        in_struct,
        &mut s,
    )?;
    Ok(s)
}

fn gen_type_inner<W: fmt::Write>(
    typ: &ast::TypeName,
    in_path: &ast::Path,
    ref_type: Option<RefType>,
    env: &Env,
    library_config: &LibraryConfig,
    in_struct: bool,
    out: &mut W,
) -> fmt::Result {
    let mut handled_ref = false;
    match typ {
        ast::TypeName::Named(path_type) | ast::TypeName::SelfType(path_type) => {
            match path_type.resolve(in_path, env) {
                ast::CustomType::Opaque(opaque) => {
                    match ref_type {
                        None => panic!("Cannot pass opaque structs as values"),
                        Some(RefType::Owned) => write!(out, "{}", opaque.name)?,
                        Some(RefType::BorrowedNullable) => write!(out, "{}*", opaque.name)?,
                        Some(RefType::BorrowedNonNull) => write!(out, "{}&", opaque.name)?,
                    }
                    handled_ref = true;
                }

                ast::CustomType::Struct(strct) => {
                    write!(out, "{}", strct.name)?;
                }

                ast::CustomType::Enum(enm) => {
                    write!(out, "{}", enm.name)?;
                }
                &_ => unreachable!("unknown AST/HIR variant"),
            }
        }

        ast::TypeName::Box(underlying) => {
            gen_type_inner(
                underlying.as_ref(),
                in_path,
                Some(RefType::Owned),
                env,
                library_config,
                in_struct,
                out,
            )?;
        }

        ast::TypeName::Reference(_, mutability, underlying) => {
            if mutability.is_immutable() && !in_struct {
                write!(out, "const ")?;
            }
            gen_type_inner(
                underlying.as_ref(),
                in_path,
                Some(RefType::BorrowedNonNull),
                env,
                library_config,
                in_struct,
                out,
            )?;
        }

        ast::TypeName::Option(underlying) => match underlying.as_ref() {
            ast::TypeName::Box(_) => {
                write!(out, "{}<", library_config.optional.expr)?;
                gen_type_inner(
                    underlying.as_ref(),
                    in_path,
                    ref_type,
                    env,
                    library_config,
                    in_struct,
                    out,
                )?;
                write!(out, ">")?;
            }

            ast::TypeName::Reference(_, mutability, underlying) => {
                if mutability.is_immutable() && !in_struct {
                    write!(out, "const ")?;
                }
                gen_type_inner(
                    underlying.as_ref(),
                    in_path,
                    Some(RefType::BorrowedNullable),
                    env,
                    library_config,
                    in_struct,
                    out,
                )?;
            }

            _ => unreachable!("unknown AST/HIR variant"),
        },

        ast::TypeName::Result(ok, err, _) => {
            write!(out, "diplomat::result<")?;
            if ok.is_zst() {
                write!(out, "std::monostate")?;
            } else {
                gen_type_inner(ok, in_path, ref_type, env, library_config, in_struct, out)?;
            }

            write!(out, ", ")?;
            if err.is_zst() {
                write!(out, "std::monostate")?;
            } else {
                gen_type_inner(err, in_path, ref_type, env, library_config, in_struct, out)?;
            }
            write!(out, ">")?;
        }

        ast::TypeName::Primitive(prim) => {
            write!(out, "{}", crate::c::types::c_type_for_prim(prim))?;
        }

        ast::TypeName::Writeable => {
            write!(out, "capi::DiplomatWriteable")?;
        }

        ast::TypeName::StrReference(
            _,
            ast::StringEncoding::UnvalidatedUtf8 | ast::StringEncoding::Utf8,
        ) => {
            let maybe_const = if in_struct { "" } else { "const " };
            write!(out, "{maybe_const}{}", library_config.string_view.expr)?;
        }

        ast::TypeName::StrReference(_, ast::StringEncoding::UnvalidatedUtf16) => {
            let maybe_const = if in_struct { "" } else { "const " };
            write!(out, "{maybe_const}{}", library_config.string_view16.expr)?;
        }

        ast::TypeName::PrimitiveSlice(_, ast::Mutability::Mutable, prim) => {
            write!(
                out,
                "{}<const {}>",
                library_config.span.expr,
                crate::c::types::c_type_for_prim(prim)
            )?;
        }

        ast::TypeName::PrimitiveSlice(_, ast::Mutability::Immutable, prim) => {
            write!(
                out,
                "const {}<const {}>",
                library_config.span.expr,
                crate::c::types::c_type_for_prim(prim)
            )?;
        }

        ast::TypeName::Unit => {
            write!(out, "void")?;
        }
        &_ => unreachable!("unknown AST/HIR variant"),
    }

    if !handled_ref {
        match ref_type {
            None => (),
            Some(RefType::Owned) | Some(RefType::BorrowedNullable) => write!(out, "*")?,
            Some(RefType::BorrowedNonNull) => write!(out, "&")?,
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_pointer_types() {
        test_file! {
            #[diplomat::bridge]
            mod ffi {
                #[diplomat::opaque]
                struct MyOpaqueStruct(UnknownType);

                struct MyStruct {
                    a: Box<MyOpaqueStruct>,
                }

                impl MyStruct {
                    pub fn new(foo: &MyOpaqueStruct, bar: &mut MyOpaqueStruct) -> MyStruct {
                        unimplemented!()
                    }
                }
            }
        }
    }

    #[test]
    fn test_option_types() {
        test_file! {
            #[diplomat::bridge]
            mod ffi {
                #[diplomat::opaque]
                struct MyOpaqueStruct(UnknownType);

                struct MyStruct {
                    a: Option<Box<MyOpaqueStruct>>,
                }
            }
        }
    }

    #[test]
    fn test_option_types_using_library_config() {
        test_file_using_library_config! {
            #[diplomat::bridge]
            mod ffi {
                #[diplomat::opaque]
                struct MyOpaqueStruct(UnknownType);

                struct MyStruct {
                    a: Option<Box<MyOpaqueStruct>>,
                }
            }
        }
    }

    #[test]
    fn test_result_types() {
        test_file! {
            #[diplomat::bridge]
            mod ffi {
                #[diplomat::opaque]
                struct MyOpaqueStruct(UnknownType);

                struct MyStruct {
                    a: DiplomatResult<Box<MyOpaqueStruct>, u8>,
                }

                impl MyStruct {
                    pub fn new() -> Result<MyStruct, u8> {
                        unimplemented!()
                    }
                }
            }
        }
    }

    #[test]
    fn test_string_reference() {
        test_file! {
            #[diplomat::bridge]
            mod ffi {
                struct MyStruct;

                impl MyStruct {
                    pub fn new(v: &DiplomatStr) -> MyStruct {
                        unimplemented!()
                    }
                }
            }
        }
    }

    #[test]
    fn test_writeable_out() {
        test_file! {
            #[diplomat::bridge]
            mod ffi {
                struct MyStruct;

                impl MyStruct {
                    pub fn write(&self, to: &mut DiplomatWriteable) {
                        unimplemented!()
                    }
                }
            }
        }
    }

    #[test]
    fn test_unit_type() {
        test_file! {
            #[diplomat::bridge]
            mod ffi {
                struct MyStruct;

                impl MyStruct {
                    pub fn something(&self) -> () {
                        unimplemented!()
                    }
                }
            }
        }
    }
}
