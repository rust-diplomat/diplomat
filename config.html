<!DOCTYPE HTML>
<html lang="en" class="light sidebar-visible" dir="ltr">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>Configuration - The Diplomat Book</title>


        <!-- Custom HTML head -->

        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff">

        <link rel="icon" href="favicon.svg">
        <link rel="shortcut icon" href="favicon.png">
        <link rel="stylesheet" href="css/variables.css">
        <link rel="stylesheet" href="css/general.css">
        <link rel="stylesheet" href="css/chrome.css">
        <link rel="stylesheet" href="css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" id="highlight-css" href="highlight.css">
        <link rel="stylesheet" id="tomorrow-night-css" href="tomorrow-night.css">
        <link rel="stylesheet" id="ayu-highlight-css" href="ayu-highlight.css">

        <!-- Custom theme stylesheets -->


        <!-- Provide site root and default themes to javascript -->
        <script>
            const path_to_root = "";
            const default_light_theme = "light";
            const default_dark_theme = "navy";
        </script>
        <!-- Start loading toc.js asap -->
        <script src="toc.js"></script>
    </head>
    <body>
    <div id="mdbook-help-container">
        <div id="mdbook-help-popup">
            <h2 class="mdbook-help-title">Keyboard shortcuts</h2>
            <div>
                <p>Press <kbd>←</kbd> or <kbd>→</kbd> to navigate between chapters</p>
                <p>Press <kbd>S</kbd> or <kbd>/</kbd> to search in the book</p>
                <p>Press <kbd>?</kbd> to show this help</p>
                <p>Press <kbd>Esc</kbd> to hide this help</p>
            </div>
        </div>
    </div>
    <div id="body-container">
        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                let theme = localStorage.getItem('mdbook-theme');
                let sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            const default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? default_dark_theme : default_light_theme;
            let theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            const html = document.documentElement;
            html.classList.remove('light')
            html.classList.add(theme);
            html.classList.add("js");
        </script>

        <input type="checkbox" id="sidebar-toggle-anchor" class="hidden">

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            let sidebar = null;
            const sidebar_toggle = document.getElementById("sidebar-toggle-anchor");
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
            }
            sidebar_toggle.checked = sidebar === 'visible';
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <!-- populated by js -->
            <mdbook-sidebar-scrollbox class="sidebar-scrollbox"></mdbook-sidebar-scrollbox>
            <noscript>
                <iframe class="sidebar-iframe-outer" src="toc.html"></iframe>
            </noscript>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle">
                <div class="sidebar-resize-indicator"></div>
            </div>
        </nav>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky">
                    <div class="left-buttons">
                        <label id="sidebar-toggle" class="icon-button" for="sidebar-toggle-anchor" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </label>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="default_theme">Auto</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search (`/`)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="/ s" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">The Diplomat Book</h1>

                    <div class="right-buttons">
                        <a href="print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <h1 id="configuring-diplomat"><a class="header" href="#configuring-diplomat">Configuring Diplomat</a></h1>
<p>Some Diplomat backends have configurable parameters to change the behavior of their output.</p>
<p>For instance, Kotlin requires a <code>domain</code> parameter and <code>lib_name</code> parameter. These two parameters are used to generate a package name and folder structure on generation.</p>
<h2 id="configuration-structures"><a class="header" href="#configuration-structures">Configuration Structures</a></h2>
<p>Documentation on what configuration options are available in <a href="https://github.com/rust-diplomat/diplomat/tree/main/tool/src/config.rs"><code>config.rs</code> on GitHub</a>.</p>
<p>Note that every configuration option uses <code>snake_case</code> for consistency, and that the <code>SharedConfig</code> struct is flattened. So for setting <code>--config shared_config.lib_name="some_value"</code>, this would instead be <code>--config lib_name="some_value"</code>.</p>
<h2 id="configuration-interfaces"><a class="header" href="#configuration-interfaces">Configuration Interfaces</a></h2>
<p>Configuration information can be set in four ways:</p>
<h3 id="configtoml"><a class="header" href="#configtoml"><code>config.toml</code></a></h3>
<p>By default, Diplomat scans for a <code>config.toml</code> in the folder where <code>diplomat_tool</code> is being run. You can change the location of this folder with the <code>--config_file</code> parameter.</p>
<p>The structure of <code>config.toml</code> is as follows:</p>
<pre><code class="language-toml"># Top level table specifies Shared Config settings that apply to all backends:
lib-name = "MyLibrary"

[kotlin]
# Individual tables can override Shared Config settings:
lib-name = "LibraryNameOverride"
# Along with backend specific settings:
domain = "org.myOrganization"

[demo_gen]
explicit-generation = true

[other-library-name]
some-value = 100
</code></pre>
<h3 id="diplomat-tool-cli"><a class="header" href="#diplomat-tool-cli"><code>diplomat-tool</code> CLI</a></h3>
<p>When running <code>diplomat-tool</code>, you may pass in the <code>--config</code> flag for each option you wish to set:</p>
<pre><code>./diplomat-tool kotlin ./kotlin-folder --config lib_name="MyLibrary" --config kotlin.domain = "org.myOrganization"
</code></pre>
<p><code>diplomat-tool</code> flags take priority over <code>config.toml</code>.</p>
<h3 id="diplomat_toolgen"><a class="header" href="#diplomat_toolgen"><code>diplomat_tool::gen</code></a></h3>
<p>If you call <code>diplomat_tool::gen</code> manually, then you have the option of setting configuration yourself, with the <code>diplomat_tool::config::Config</code> struct.</p>
<p>See <a href="#backend-structures">Backend Structures</a> for more on these structures.</p>
<h3 id="diplomatconfig"><a class="header" href="#diplomatconfig"><code>#[diplomat::config(...)]</code></a></h3>
<p>In <code>lib.rs</code>, any top-level <code>mod</code>ule, <code>struct</code>, or <code>impl</code> block can use the <code>#[diplomat::config]</code> attribute:</p>
<pre><pre class="playground"><code class="language-rust"><span class="boring">#![allow(unused)]
</span><span class="boring">fn main() {
</span>#[diplomat::config(lib_name="MyLibrary")]
struct SomeConfig;

#[diplomat::config(kotlin.domain="org.myOrganization")]
mod kotlin_specific_mod;

#[diplomat::config(...)]
impl SomeConfig {

}
<span class="boring">}</span></code></pre></pre>
<p>Due to a quirk of how Diplomat reads these attributes, <code>#[diplomat::config]</code> has priority over all other methods of setting configuration.</p>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                            <a rel="prev" href="basics.html" class="mobile-nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                                <i class="fa fa-angle-left"></i>
                            </a>

                            <a rel="next prefetch" href="types.html" class="mobile-nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                                <i class="fa fa-angle-right"></i>
                            </a>

                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
                    <a rel="prev" href="basics.html" class="nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                        <i class="fa fa-angle-left"></i>
                    </a>

                    <a rel="next prefetch" href="types.html" class="nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                        <i class="fa fa-angle-right"></i>
                    </a>
            </nav>

        </div>




        <script>
            window.playground_copyable = true;
        </script>


        <script src="elasticlunr.min.js"></script>
        <script src="mark.min.js"></script>
        <script src="searcher.js"></script>

        <script src="clipboard.min.js"></script>
        <script src="highlight.js"></script>
        <script src="book.js"></script>

        <!-- Custom JS scripts -->


    </div>
    </body>
</html>
