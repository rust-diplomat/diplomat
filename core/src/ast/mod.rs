/// As part of the macro expansion and code generation process, Diplomat
/// generates a simplified version of the Rust AST that captures special
/// types such as opaque structs, [`Box`], and [`Result`] with utilities
/// for handling such types.
pub mod attrs;
pub use attrs::Attrs;

mod methods;
pub use methods::{BorrowedParams, Method, Param, SelfParam};

mod modules;
pub use modules::{File, Module};

mod structs;
pub use structs::{OpaqueStruct, Struct};

mod enums;
pub use enums::Enum;

mod types;
pub use types::{
    CustomType, LifetimeOrigin, ModSymbol, Mutability, PathType, PrimitiveType, TypeName,
};

mod lifetimes;
pub use lifetimes::{Lifetime, LifetimeEnv, LifetimeTransitivity, NamedLifetime};

mod paths;
pub use paths::Path;

mod idents;
pub use idents::Ident;

mod docs;
pub use docs::{DocType, Docs, DocsUrlGenerator, MarkdownStyle, RustLink, RustLinkDisplay};

mod validity;
pub use validity::ValidityError;
