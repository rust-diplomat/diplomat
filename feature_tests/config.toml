lib-name = "This Will Be Ignored For Testing Purposes (Check src/lib.rs)"
unsafe_references_in_callbacks = true

[kotlin]
domain = "dev.diplomattest"

[demo-gen]
relative-js-path = "../../js/api/"