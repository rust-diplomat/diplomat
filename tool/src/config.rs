use std::{collections::HashMap, path::Path, str};

use quote::ToTokens;
use serde::{Deserialize, Serialize};
use syn::{
    parse::{Parse, ParseStream},
    Expr, Ident, Token,
};
use toml::{value::Table, Value};

use crate::{demo_gen::DemoConfig, kotlin::KotlinConfig};

#[derive(Clone, Default, Debug, Serialize, Deserialize)]
pub struct SharedConfig {
    pub lib_name: Option<String>,
}

impl SharedConfig {
    /// Quick and dirty way to tell [`set_overrides`] whether or not to copy an override from a specific language over.
    pub fn overrides_shared(name: &str) -> bool {
        // Expect the first item in the iterator to be the name of the language, so we eliminate that:
        let name: String = name.split(".").skip(1).collect();
        matches!(name.as_str(), "lib_name")
    }

    pub fn set(&mut self, key: &str, value: Value) {
        if key == "lib_name" && value.is_str() {
            self.lib_name = value.as_str().map(|v| v.to_string());
        }
    }
}

#[derive(Clone, Default, Debug, Serialize, Deserialize)]
pub struct Config {
    #[serde(flatten)]
    pub shared_config: SharedConfig,
    #[serde(rename = "kotlin")]
    pub kotlin_config: KotlinConfig,
    #[serde(rename = "demo_gen")]
    pub demo_gen_config: DemoConfig,
    /// Any language can override what's in [`SharedConfig`]. This is a structure that holds information about those specific overrides. [`Config`] will update [`SharedConfig`] based on the current language.
    #[serde(skip)]
    pub language_overrides: HashMap<String, Value>,
}

impl Config {
    pub fn set(&mut self, key: &str, value: Value) {
        if key.starts_with("kotlin.") {
            if SharedConfig::overrides_shared(key) {
                self.language_overrides.insert(key.to_string(), value);
            } else {
                self.kotlin_config.set(&key.replace("kotlin.", ""), value);
            }
        } else if key.starts_with("demo_gen.") {
            if SharedConfig::overrides_shared(key) {
                self.language_overrides.insert(key.to_string(), value);
            } else {
                self.demo_gen_config
                    .set(&key.replace("demo_gen.", ""), value);
            }
        } else if key.starts_with("nanobind.") {
            if SharedConfig::overrides_shared(key) {
                self.language_overrides.insert(key.to_string(), value);
            } // nanobind doesn't have any other config setting
        } else {
            self.shared_config.set(key, value)
        }
    }

    pub fn get_overridden(self, target_language: &str) -> Self {
        let mut out = self.clone();

        // Look for a match of language_name.some_value in a potential key.
        let m = format!("{}.", target_language);
        for (k, v) in out.language_overrides.iter() {
            if k.starts_with(&m) {
                out.shared_config.set(&k.replace(&m, ""), v.clone());
            }
        }
        out
    }

    /// Given a filepath, read TOML formatted config settings from it (and modify the current Config struct from the read)
    pub fn read_file(&mut self, path: &Path) -> Result<(), String> {
        let config_table: Table = if path.exists() {
            let file_buf = std::fs::read(path).map_err(|e| e.to_string())?;
            let s = str::from_utf8(&file_buf).map_err(|_| "Config file is not UTF8".to_string())?;
            toml::from_str(s).map_err(|_| "Config file is not valid TOML".to_string())?
        } else {
            Table::default()
        };

        for (key, value) in config_table {
            // Quick way to take config.toml from kebab to snake case.
            // This technically means that someone could also just as easily do CamelCase and have it translated,
            // but I'm not sure I want to bother writing validation code for such a scenario.
            let key = heck::AsSnakeCase(key).to_string();
            if let toml::Value::Table(t) = value {
                for (subkey, subvalue) in t {
                    let subkey = heck::AsSnakeCase(subkey).to_string();
                    self.set(&format!("{}.{}", key, subkey), subvalue);
                }
            } else {
                self.set(&key, value);
            }
        }
        Ok(())
    }

    /// Given a vector of strings with the format `config.setting = value`, modify the `Config` struct appropriately.
    pub fn read_cli_settings(&mut self, settings: Vec<String>) {
        for c in settings {
            let split = c.split_once("=");
            if let Some((key, value)) = split {
                self.set(key, toml_value_from_str(value));
            } else {
                eprintln!("Could not read {c}, expected =");
            }
        }
    }
}

pub fn toml_value_from_str(string: &str) -> toml::Value {
    let try_parse = toml::from_str::<toml::Value>(string);

    // If there's an error parsing (because clap will not parse quotes, for example), we just treat what we're passed as a string:
    if let Ok(out) = try_parse {
        out
    } else {
        toml::Value::String(string.to_string())
    }
}

#[derive(Clone, PartialEq, Eq, Hash, Debug, Serialize)]
#[non_exhaustive]
pub(crate) struct DiplomatBackendConfigAttr {
    pub key_value_pairs: Vec<DiplomatBackendConfigKeyValue>,
}

impl Parse for DiplomatBackendConfigAttr {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let list = input.parse_terminated(DiplomatBackendConfigKeyValue::parse, Token![,])?;
        let vec = list.into_iter().collect();
        Ok(Self {
            key_value_pairs: vec,
        })
    }
}

#[derive(Clone, PartialEq, Eq, Hash, Debug, Serialize)]
#[non_exhaustive]
pub(crate) struct DiplomatBackendConfigKeyValue {
    pub key: String,
    pub value: String,
}

impl Parse for DiplomatBackendConfigKeyValue {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let mut key_str: Vec<String> = Vec::new();

        loop {
            let i: Ident = input.parse()?;

            key_str.push(i.to_string());

            if input.peek(Token![.]) {
                let _period: Token![.] = input.parse()?;
            } else {
                break;
            }
        }

        let _equals: Token![=] = input.parse()?;

        let val_expr: Expr = input.parse()?;

        let value = val_expr.to_token_stream().to_string();

        Ok(Self {
            key: key_str.join("."),
            value,
        })
    }
}

pub(crate) fn find_top_level_attr(module_items: Vec<syn::Item>) -> Vec<DiplomatBackendConfigAttr> {
    let path = syn::parse_str("diplomat::config").unwrap();

    let attrs = module_items
        .iter()
        .filter_map(|i| match i {
            syn::Item::Struct(s) => Some(s.attrs.clone()),
            syn::Item::Impl(i) => Some(i.attrs.clone()),
            syn::Item::Mod(m) => Some(m.attrs.clone()),
            _ => None,
        })
        .filter_map(|attrs| {
            let attributes_vec = attrs
                .iter()
                .filter_map(|attribute| {
                    if attribute.path() == &path {
                        Some(
                            syn::parse2::<DiplomatBackendConfigAttr>(
                                attribute
                                    .parse_args()
                                    .expect("Failed to parse malformed diplomat::config"),
                            )
                            .expect("Could not parse DiplomatBackendConfig attribute."),
                        )
                    } else {
                        None
                    }
                })
                .collect::<Vec<_>>();

            if !attributes_vec.is_empty() {
                Some(attributes_vec)
            } else {
                None
            }
        });

    let mut out_config = Vec::new();

    for mut a in attrs {
        out_config.append(&mut a);
    }

    out_config
}
