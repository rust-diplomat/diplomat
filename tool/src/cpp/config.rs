use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub struct CallableLibraryType {
    pub name: String,
    pub expr: String,
    pub is_call: bool,
}

#[derive(Debug, Deserialize)]
pub struct LibraryType {
    pub name: String,
    pub expr: String,
}

#[derive(Debug, Deserialize)]
pub struct LibraryConfig {
    pub headers: Vec<String>,
    pub nullopt: CallableLibraryType,
    pub optional: LibraryType,
    pub someopt: CallableLibraryType,
    pub span: LibraryType,
    pub string_view: LibraryType,
    pub string_view16: LibraryType,
    pub unique_ptr: LibraryType,
}

impl LibraryConfig {
    pub fn default() -> LibraryConfig {
        LibraryConfig {
            headers: vec!["#include <optional>".into()],
            nullopt: CallableLibraryType {
                name: "std::nullopt".into(),
                expr: "std::nullopt".into(),
                is_call: false,
            },
            optional: LibraryType {
                name: "std::optional".into(),
                expr: "std::optional".into(),
            },
            someopt: CallableLibraryType {
                name: "".into(),
                expr: "".into(),
                is_call: false,
            },
            span: LibraryType {
                // This is std::span compatible, and can be replaced with std::span
                // if targeting C++20. Internally the header will `using span = std::span`
                // anyway
                name: "diplomat::span".into(),
                expr: "diplomat::span".into(),
            },
            string_view: LibraryType {
                name: "std::string_view".into(),
                expr: "std::string_view".into(),
            },
            string_view16: LibraryType {
                name: "std::u16string_view".into(),
                expr: "std::u16string_view".into(),
            },
            unique_ptr: LibraryType {
                name: "std::unique_ptr".into(),
                expr: "std::unique_ptr".into(),
            },
        }
    }
}
