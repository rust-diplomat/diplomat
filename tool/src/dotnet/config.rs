use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub struct LibraryConfig {
    pub namespace: String,
    pub native_lib: String,
    #[serde(default)]
    pub usings: Vec<String>,
    #[serde(default)]
    pub exceptions: ExceptionsConfig,
    #[serde(default)]
    pub properties: PropertiesConfig,
}

impl LibraryConfig {
    pub fn default() -> LibraryConfig {
        LibraryConfig {
            namespace: "Interop".to_owned(),
            native_lib: "rust".to_owned(),
            usings: Vec::new(),
            exceptions: ExceptionsConfig::default(),
            properties: PropertiesConfig::default(),
        }
    }
}

#[derive(Debug, Deserialize, Default)]
pub struct PropertiesConfig {
    #[serde(default)]
    pub setters_prefix: Option<String>,
    #[serde(default)]
    pub getters_prefix: Option<String>,
}

#[derive(Debug, Deserialize, Default)]
pub struct ExceptionsConfig {
    #[serde(default)]
    pub trim_suffix: String,
    #[serde(default)]
    pub error_message_method: Option<String>,
}
