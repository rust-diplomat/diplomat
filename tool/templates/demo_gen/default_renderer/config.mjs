// Don't really love this, but it works for now.
import cfg from "../diplomat.config.mjs"
export default cfg;