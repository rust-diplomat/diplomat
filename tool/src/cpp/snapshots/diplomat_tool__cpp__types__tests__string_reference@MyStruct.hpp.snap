---
source: tool/src/cpp/types.rs
expression: out_texts.get(out).unwrap()

---
#ifndef MyStruct_HPP
#define MyStruct_HPP
#include <stdint.h>
#include <stddef.h>
#include <stdbool.h>
#include <algorithm>
#include <memory>
#include <variant>
#include <optional>
#include "diplomat_runtime.hpp"

namespace capi {
#include "MyStruct.h"
}

struct MyStruct;

/**
 * A destruction policy for using MyStruct with std::unique_ptr.
 */
struct MyStructDeleter {
  void operator()(capi::MyStruct* l) const noexcept {
    capi::MyStruct_destroy(l);
  }
};
struct MyStruct {
 public:
  static MyStruct new_(const std::string_view v);
};


inline MyStruct MyStruct::new_(const std::string_view v) {
  capi::MyStruct diplomat_raw_struct_out_value = capi::MyStruct_new(v.data(), v.size());
  return MyStruct{  };
}
#endif

